ext/random/random.c

```diff
@@ -0,0 +1,871 @@
+{
+	uint32_t idx, j, n_elems, n_left;
+	Bucket *p, temp;
+	HashTable *hash;
+	zend_long rnd_idx;
+
+	n_elems = zend_hash_num_elements(Z_ARRVAL_P(array));
+
+	if (n_elems < 1) {
+		return;
+	}
+
+	hash = Z_ARRVAL_P(array);
+	n_left = n_elems;
+
+	if (EXPECTED(!HT_HAS_ITERATORS(hash))) {
+		if (hash->nNumUsed != hash->nNumOfElements) {
+			for (j = 0, idx = 0; idx < hash->nNumUsed; idx++) {
+				p = hash->arData + idx;
+				if (Z_TYPE(p->val) == IS_UNDEF) continue;
+				if (j != idx) {
+					hash->arData[j] = *p;
+				}
+				j++;
+			}
+		}
+		while (--n_left) {
+			rnd_idx = php_random_range(random, 0, n_left);
+			if (rnd_idx != n_left) {
+				temp = hash->arData[n_left];
+				hash->arData[n_left] = hash->arData[rnd_idx];
+				hash->arData[rnd_idx] = temp;
+			}
+		}
+	} else {
+		uint32_t iter_pos = zend_hash_iterators_lower_pos(hash, 0);
+
+		if (hash->nNumUsed != hash->nNumOfElements) {
+			for (j = 0, idx = 0; idx < hash->nNumUsed; idx++) {
+				p = hash->arData + idx;
+				if (Z_TYPE(p->val) == IS_UNDEF) continue;
+				if (j != idx) {
+					hash->arData[j] = *p;
+					if (idx == iter_pos) {
+						zend_hash_iterators_update(hash, idx, j);
+						iter_pos = zend_hash_iterators_lower_pos(hash, iter_pos + 1);
+					}
+				}
+				j++;
+			}
+		}
+		while (--n_left) {
+			rnd_idx = php_random_range(random, 0, n_left);
+			if (rnd_idx != n_left) {
+				temp = hash->arData[n_left];
+				hash->arData[n_left] = hash->arData[rnd_idx];
+				hash->arData[rnd_idx] = temp;
+				zend_hash_iterators_update(hash, (uint32_t)rnd_idx, n_left);
+			}
+		}
+	}
+	hash->nNumUsed = n_elems;
+	hash->nInternalPointer = 0;
+
+	for (j = 0; j < n_elems; j++) {
+		p = hash->arData + j;
+		if (p->key) {
+			zend_string_release_ex(p->key, 0);
+		}
+		p->h = j;
+		p->key = NULL;
+	}
+	hash->nNextFreeElement = n_elems;
+	if (!(HT_FLAGS(hash) & HASH_FLAG_PACKED)) {
+		zend_hash_to_packed(hash);
+	}
+}
+
+void php_random_string_shuffle(php_random *random, char *str, zend_long len)
+{
+	zend_long n_elems, rnd_idx, n_left;
+	char temp;
+	n_elems = len;
+
+	if (n_elems <= 1) {
+		return;
+	}
+
+	n_left = n_elems;
+
+	while (--n_left) {
+		rnd_idx = php_random_range(random, 0, n_left);
+		if (rnd_idx != n_left) {
+			temp = str[n_left];
+			str[n_left] = str[rnd_idx];
+			str[rnd_idx] = temp;
+		}
+	}
+}
+
+static inline zend_object *php_random_rng_common_init(zend_class_entry *ce, const php_random_rng_algo *algo, const zend_object_handlers *handlers) {
+	php_random_rng *rng = zend_object_alloc(sizeof(php_random_rng), ce);
+
+	zend_object_std_init(&rng->std, ce);
+	object_properties_init(&rng->std, ce);
+	
+	rng->algo = algo;
+	if (rng->algo->state_size > 0) {
+		rng->state = ecalloc(1, algo->state_size);
+	}
+	rng->std.handlers = handlers;
+
+	return &rng->std;
+}
+
+static void php_random_rng_common_free_obj(zend_object *object) {
+	php_random_rng *rng = php_random_rng_from_obj(object);
+
+	if (rng->state) {
+		efree(rng->state);
```

Comment:
Could set rng->state to NULL just in case - I think the object can be referenced even after `__destruct` is called.

I don't think that's the issue, though

Reply:
The object can't be referenced after free_obj, only dtor_obj.

Reply:
OK, I'll leave this as is.